libs/community/langchain_community/chat_message_histories/falkordb.py

```diff
@@ -0,0 +1,228 @@
+                )
+
+        self.database_name = database
+        self._database = self._driver.select_graph(database)
+        self._session_id = session_id
+        self._node_label = node_label
+        self._window = window
+
+        self._database.query(
+            f"MERGE (s:`{self._node_label}` {{id:$session_id}})",
+            {"session_id": self._session_id},
+        )
+
+    def _process_records(self, records: list) -> list:
+        """Process the records from FalkorDB and convert them into BaseMessage objects.
+
+        Args:
+            records (list): The raw records fetched from the FalkorDB query.
+
+        Returns:
+            list: A list of `BaseMessage` objects.
+        """
+        messages = []
+
+        for record in records:
+            content = record[0].get("data", {}).get("content", "")
+            message_type = record[0].get("type", "").lower()
+
+            if message_type == "human":
+                messages.append(
+                    HumanMessage(
+                        content=content, additional_kwargs={}, response_metadata={}
+                    )
+                )
+            elif message_type == "ai":
+                messages.append(
+                    AIMessage(
+                        content=content, additional_kwargs={}, response_metadata={}
+                    )
+                )
+            else:
+                print(f"Unknown message type: {message_type}")
+
+        return messages
+
+    @property
+    def messages(self) -> List[BaseMessage]:
+        """Retrieve the messages from FalkorDB for the session.
+
+        Returns:
+            List[BaseMessage]: A list of messages in the current session.
+        """
+        query = (
+            f"MATCH (s:`{self._node_label}`)-[:LAST_MESSAGE]->(last_message) "
+            "WHERE s.id = $session_id MATCH p=(last_message)<-[:NEXT*0.."
+            f"{self._window*2}]-() WITH p, length(p) AS length "
+            "ORDER BY length DESC LIMIT 1 UNWIND reverse(nodes(p)) AS node "
+            "RETURN {data:{content: node.content}, type:node.type} AS result"
+        )
+
+        records = self._database.query(
+            query, {"session_id": self._session_id}
+        ).result_set
+
+        messages = self._process_records(records)
+        return messages
+
+    @messages.setter
+    def messages(self, messages: List[BaseMessage]) -> None:
+        """Block direct assignment to 'messages' to prevent misuse."""
+        raise NotImplementedError(
+            "Direct assignment to 'messages' is not allowed."
+            " Use the 'add_message' method instead."
+        )
+
+    def add_message(self, message: BaseMessage) -> None:
+        """Append a message to the session in FalkorDB.
+
+        Args:
+            message (BaseMessage): The message object to add to the session.
+        """
+        create_query = (
+            f"MATCH (s:`{self._node_label}`) WHERE s.id = $session_id "
```

Comment:
do you create index in the possible labels on the id property? it will improve performance

Reply:
I do now
```
try:
    self._database.create_node_vector_index(
        f"{self._node_label}", "id", dim=5, similarity_function="cosine"
    )
except Exception as e:
    if "already indexed" in e:
        raise ValueError(f"{self._node_label} has already been indexed")
```